src/core/server/base.js

```diff
@@ -54,17 +53,53 @@ function getNoScriptStyles({ appName }) {
   return undefined;
 }
 
-function showErrorPage(res, status) {
-  let adjustedStatus = status;
-  let error = getErrorMsg(adjustedStatus);
-  if (!error) {
-    adjustedStatus = 500;
-    error = getErrorMsg(adjustedStatus);
+const appName = config.get('appName');
+
+function showErrorPage(res, createStore, status, error = {}) {
+  const store = createStore();
+
+  // Get SRI for deployed services only.
+  const sriData = (isDeployed) ? JSON.parse(
+    fs.readFileSync(path.join(config.get('basePath'), 'dist/sri.json'))
+  ) : {};
+
+  // Check the lang supplied by res.locals.lang for validity
+  // or fall-back to the default.
+  const lang = isValidLang(res.locals.lang) ?
+    res.locals.lang : config.get('defaultLang');
+  const dir = getDirection(lang);
+  store.dispatch(setLang(lang));
+  if (res.locals.clientApp) {
+    store.dispatch(setClientApp(res.locals.clientApp));
+  } else {
+    log.warn(`No clientApp for this ${status} error page`);
   }
-  return res.status(adjustedStatus).end(error);
-}
 
-const appName = config.get('appName');
+  const apiError = createApiError({
+    apiURL: res.url,
+    response: { status },
+  });
+  store.dispatch(loadFail('ServerBase', { ...apiError, ...error }));
+
+  const pageProps = {
+    appName,
+    assets: webpackIsomorphicTools.assets(),
+    htmlLang: lang,
+    htmlDir: dir,
+    includeSri: isDeployed,
+    noScriptStyles: '',
+    sriData,
+    store,
+    trackingEnabled: convertBoolean(config.get('trackingEnabled')),
+  };
+
+  const HTML = ReactDOM.renderToString(
+    <ServerHtml {...pageProps} />);
+  const httpStatus = NestedStatus.rewind();
```

Comment:
Fair enough. I liked the API a lot but I didn't dig too deep on how they worked. I'd be fine converting to Redux in a future patch.